src/LinalgMsg.chpl

```diff
@@ -181,71 +160,89 @@ module LinalgMsg {
         and then removing the degenerate method. This procedure expects that such a
         transformation is handled on the server side.
     */
-    @arkouda.registerND
-    proc matMulMsg(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab, param nd: int): MsgTuple throws {
-        if nd < 2 {
-            const errorMsg = "Matrix multiplication with arrays of dimension < 2 is not supported";
-            linalgLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
-            return new MsgTuple(errorMsg, MsgType.ERROR);
-        }
 
-        const x1Name = msgArgs.getValueOf("x1"),
-              x2Name = msgArgs.getValueOf("x2");
+    @arkouda.instantiateAndRegister
+    proc matmul(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab, param array_nd: int): MsgTuple throws 
+        where array_nd >= 2 {
+
+        // Get the left and right arguments.
+
+        const x1Name = msgArgs["x1"].toScalar(string),
+              x2Name = msgArgs["x2"].toScalar(string);
 
         var x1G: borrowed GenSymEntry = getGenericTypedArrayEntry(x1Name, st),
             x2G: borrowed GenSymEntry = getGenericTypedArrayEntry(x2Name, st);
 
-        const rname = st.nextName();
-
         linalgLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
-            "cmd: %s dtype1: %s dtype2: %s rname: %s".format(
-            cmd,dtype2str(x1G.dtype),dtype2str(x2G.dtype),rname));
-
-        proc doMatMult(type x1Type, type x2Type, type resultType): MsgTuple throws {
-            var x1E = toSymEntry(x1G, x1Type, nd),
-                x2E = toSymEntry(x2G, x2Type, nd);
-
-            const (valid, outDims, err) = assertValidDims(x1E, x2E);
-            if !valid then return err;
-
-            var eOut = st.addEntry(rname, (...outDims), resultType);
-
-            const x1 = x1E.a : resultType,
-                  x2 = x2E.a : resultType;
-
-            if nd == 2
-                then matMult(x1, x2, eOut.a);
-                else batchedMatMult(x1, x2, eOut.a);
+            "cmd: %s dtype1: %s dtype2: %s".format(
+            cmd,dtype2str(x1G.dtype),dtype2str(x2G.dtype)));
 
-            const repMsg = "created " + st.attrib(rname);
-            linalgLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),repMsg);
-            return new MsgTuple(repMsg, MsgType.NORMAL);
-        }
+        //  The select block below (and in vecdot) implements the rule:
+        //     if either input is real, the output is real
+        //     else if either input is integer, the output is integer
+        //     else if either input is uint(8), the output is uint(8)
+        //     else if both inputs are bool, the output is bool
+        //     else error condition.
 
         select (x1G.dtype, x2G.dtype) {
+            // real cases
```

Comment:
The type logic could be lifted into the signature here:

```chapel
@arkouda.instantiateAndRegister
proc matmul(cmd: ..., type array_dtype_x1, type array_dtype_x2, param array_nd: int): MsgTuple throws {
  const x1 = st[msgArgs["x1"]]: borrowed SymEntry(array_dtype_x1, array_nd);
  const x2 = st[msgArgs["x2"]]: borrowed SymEntry(array_dtype_x2, array_nd);

  type result_type = compute_result_type(array_dtype_x1, array_dtype_x2);

  ...

}

proc compute_result_type(type t1, type t2) type {
  if t1 == real || t2 == real then return real;
  if t1 == int || t2 == int then return int;
  if t1 == uin(8) || t2 == uint(8) then return uint(8);
  return bool;
}

Reply:
hey @jeremiah-corrado, I'm struggling a bit to create a function that returns a type in chpl. I did this off to the side:
```chapel
proc myHelper(retFloat:bool) {
    if retFloat {
        return real;
    }
    else {
        return int;
    }
}

type t = myHelper(true);
```
and got a compiler error `error: illegal return of type where value is expected`

Reply:
For that proc to work, the argument needs to be a `param` s.t. the if-block can be evaluated at compile time, and it needs a type "return intent":
```chapel
proc myHelper(param retFloat: bool) type { ... }
```

Reply:
ah! I tried each of these individually but never combined them

Reply:
This is a great suggestion.  I  made the change.